Matrix operations

RafaelRMA opened the pull request:
Added: 

-matrix sum

-scalar multiplication

-matrix multiplication

WesleyCeraso commented on src/math/p_matrix_smul_f32.c:
```diff
@@ -0,0 +1,27 @@
+ *
+ */
+
+void p_matrix_smul_f32(const float *a, const float *b, float *c, int n1, int n2, int p, p_team_t team){
```
Why do b is a pointer, is it necessary?

ebadi commented:
I don't think parameters int n3, int n4 and error checking is necessary for src/math/p_matrix_sum_f32.c

RafaelRMA commented:
Why not?

RafaelRMA commented:
Actually I think it's better to use pointers.

WesleyCeraso commented:
It's slower than pass-by-value when it comes to primitive types.

http://stackoverflow.com/questions/14013139/is-it-counter-productive-to-pass-primitive-types-by-reference
http://stackoverflow.com/questions/4112914/is-it-better-to-pass-by-value-or-by-reference-for-basic-datatypes
http://stackoverflow.com/questions/10770410/reasons-to-not-pass-simple-types-by-reference

RafaelRMA commented:
So tell me... Why do the creators of PAL use pointers? They won't do this if it was slower

tcptomato commented:
For vectors it makes sense. For scalar values, not so much ( speed wise). Now it depends if you want to be consistent in the interface or not. ( inputs are always passed by pointers ).  This is a discussion that should be had.

RafaelRMA commented:
And I use less memory this way

WesleyCeraso commented:
Pointers are used when it's needed to be used. In the function p_matrix_smul_f32 the variable b is not an array, so it is not necessary to be a pointer, a pass-by-value is enough, and faster.
Using pointer to pass primitive values you are adding one more thing in the stack, using more memory!

tcptomato commented:
And prevent the compiler from doing some optimizations, because of pointer aliasing. As i said, there is a deeper discussion to be had here.

mateunho commented:
2 of 3 functions you introduced have no sense:
- `p_matrix_smul_f32` does exactly the same thing as `p_mul_f32`
- `p_matrix_sum_f32` does the same as `p_add_f32`

No matter how you name those two functions, they operate on vectors.

RafaelRMA commented:
Ah... you are talking about the smul function.yes, i agree its better to use pass-by-vallue. I will fix it. Sorry I thought you were talking about all the functions in general.

RafaelRMA commented:
@mateunho I haven't noticed this. Well, at least the p_matrix_mul_f32 was a contribution. I will fix all this things.

mateunho commented:
You can leave `p_matrix_smul_f32`, but rename it to sth like `p_smul_f32`, cause it's no connection with matrices.

RafaelRMA commented:
Ok

RafaelRMA commented:
Work done. Now i'm working on a determinant calculator function. I will do it saturday.

RafaelRMA commented:
Work done. Now i'm working on a determinant calculator function. I will do it saturday.

RafaelRMA commented:
Work done. Now i'm working on a determinant calculator function. I will do it saturday.

RafaelRMA commented:
Sometimes I can't uderstand the Github. I've modified the Makefile.am(removed the p_matrix_sum_f32.c \ line) and I got a Travis error. Then I just added a blank space in a random line and, like magic, it worked.

mateunho commented:
@RafaelRMA please read more about [coding style](https://github.com/parallella/pal/blob/master/CODING-STYLE.md) in pal and apply it to your PR.

RafaelRMA commented:
I usually care about coding style. Have a look in one of mine piece of codes:
 cout << "2"<<endl;
    for(int i=3; ind<x; i++) //i:numero a ser testado
    {
        primesqrt = ceil(sqrt(i));
        for(; n<x,prime[n]<=(int)primesqrt; n++)
        {
            if(i % prime[n] == 0)
            {
                prime[ind] = 0 ;
                break;
            }
            if(i % prime[n] !=0)
            {
                prime[ind] = i ;
            }
        }
        n=0;
        if(prime[ind] != 0)
        {
            ind++;
            cout << prime[ind-1]<<endl;
        }
    }

I supose That's what you want.

WesleyCeraso commented:
@RafaelRMA, what @mateunho tried to tell you is that you have to follow the code style applied to this project. My code style is also different from the one used in this project, but I have to follow this project's guidelines when contributing to it.
This project's code style can be found in the link @mateunho sent you, or you can look at other's code.

For example: except for function declarations, the first brace must be in the same line:
if () {
for () {
etc..

RafaelRMA commented:
I understand, I will fix it as soon as possible

RafaelRMA commented:
I hope everything is on the coding style. Sorry for this little trouble, I just want to help.

RafaelRMA commented:
What's wrong? Could you help me?

WesleyCeraso commented:
You have to look at the build log, for the first commit it shows:
  CC     p_smul_f32.lo
../../../src/math/p_smul_f32.c: In function ‘p_smul_f32’:
../../../src/math/p_smul_f32.c:29:5: error: expected declaration or statement at end of input

RafaelRMA commented:
I'll do it at home. It's impossible to do it on the smartphone.

tcptomato commented:
You're missing a closing }. Or better, remove the ones around the for.

RafaelRMA commented:
Thanks, I miss it.

tcptomato commented:
There is the "make check" target which would compile locally and show you the error faster.

RafaelRMA commented:
I didn't know this, thanks.

tcptomato commented:
Well the README skipped it in Build Instructions. I got it from the Travis script.

RafaelRMA commented:
OK... I think everything is correct now, am I right?

mateunho commented:
In my opinion there're too many arguments to `p_matrix_mul_f32`. When multiplying two matrices, they must have common dimension.
In case ` A x B = C `, there is `a_rows`, `common` and `b_cols`, eg.: `A` is 5 x __2__, `B` is __2__ x 3.
It's useless to pass four arguments of dimensions.
Passing those three is also much less error prone.

mateunho commented on src/math/p_smul_f32.c:
```diff
@@ -0,0 +1,29 @@
+ *
+ * Scalar multiplication
+ *
+ * @param a     Pointer to input matrix
```
input vector

mateunho commented on src/math/p_smul_f32.c:
```diff
@@ -0,0 +1,29 @@
+ *
+ * @param b     input scalar
+ *
+ * @param c     Pointer to output matrix
```
output vector

mateunho commented on src/math/p_smul_f32.c:
```diff
@@ -0,0 +1,29 @@
+ *
+ * @param n1    Number of lines of the matrix
+ *
+ * @param n2    Number of columns of the matrix
```
useless

mateunho commented on src/math/p_smul_f32.c:
```diff
@@ -0,0 +1,29 @@
+ *
+ * @param c     Pointer to output matrix
+ *
+ * @param n1    Number of lines of the matrix
```
vector length

WesleyCeraso commented:
I have one little thing to say. There must be a whitespace between the parenthesis and the opening brace. for ()_{

WesleyCeraso commented:
Another thing. We are not using returns. See the discussion in the end of #57

WesleyCeraso commented:
Also, the 'for' loop initial declarations ( *for (int i = 0; ...* ) are only allowed in C99 or C11 mode, I'm not sure which version we are targeting, maybe @aolofsson could tell us.

Additionally, we should use size_t (curently we are using ssize_t) for vector length.

RafaelRMA commented:
I agree with you @mateunho , but I did i t this way to prevent errors.  But I could change it, what do you think?

RafaelRMA commented:
@WesleyCeraso I knew this C version problem, but I thought it won't have problems here. To prevent errors I will declare the variables outside the for.

RafaelRMA commented:
@mateunho what exactly do you mean when you say "Additionally, we should use size_t (curently we are using ssize_t) for vector length." ?

WesleyCeraso commented:
@RafaelRMA I said that, because there does not exist negative length, so we should not allow the user to input one, but it's still in discussion in #74. For now, let the int.

Once it's all ok, don't forget to squash your commits.

RafaelRMA commented:
I got a "This pull request contains merge conflicts that must be resolved." What should I do?

WesleyCeraso commented:
In your machine, add parallella/pal as a remote
(git remote add parallella https://github.com/parallella/pal.git)

Fetch the repository
(git fetch -p)

Rebase your code
(git rebase parallella/master)

Solve the conflicts

Push again
(git push origin <your branch> -f)

------------------------------

Don't forget to squash your commits!!

------------------------------

Saudações

Wesley Ceraso Prudencio

2015-06-12 22:33 GMT-03:00 Rafael Marinho <notifications@github.com>:

> I got a "This pull request contains merge conflicts that must be
> resolved." What should I do?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/parallella/pal/pull/120#issuecomment-111651488>.
>

RafaelRMA commented:
The "git push origin <your branch> -f" failed. What sohuld I do now?

WesleyCeraso commented:
Why did it fail?

RafaelRMA commented:
Can I close this pull request and open another? Matbe it will be easier this way.

mateunho commented:
Hey, you can clone my branch: [rafael](https://github.com/mateunho/pal/tree/rafael). I've rebased your work and merged with current repo state.

RafaelRMA commented:
I've cloned your branch, but what I do now?

RafaelRMA commented:
I tried everything, now the git shell don't find any conflicts but here it stills the same.

mateunho commented:
After cloning my branch as it is, execute:
```
git push -f origin master
```
it will __overwrite__ your repo branch on github, making it mergeable with pal.

RafaelRMA commented:
I got:
remote: Permission to mateunho/pal.git denied to f03605b645198c446635432ffe0beceeadd5266c.
fatal: unable to access 'https://github.com/mateunho/pal.git/': The requested URL returned error: 403

mateunho commented:
I forgot to mention you have to add `remote`:
```
git remote add my_repo https://github.com/RafaelRMA/pal.git
```
and then you go:
```
git push -f my_repo master
```

RafaelRMA commented:
I think I did something wrong...

RafaelRMA commented:
I got
error: src refspec master does not match any.
error: failed to push some refs to 'https://github.com/RafaelRMA/pal.git'

mateunho commented:
Stackoverflow comes with help [#1](http://stackoverflow.com/questions/10568641/git-error-src-refspec-master-does-not-match-any) and [#2](http://stackoverflow.com/questions/21264738/error-src-refspec-master-does-not-match-any). One of them should apply.

RafaelRMA commented:
Now i think i did right.

RafaelRMA commented:
Finally

mateunho commented:
Nope, you didn't make it right. I have an idea. Please, add me as a collaborator to your fork of pal and I'll fix it.

RafaelRMA commented:
Thanks for helping me
